### PR TITLE
ARROW-7542: [CI][C++] Use $(sysctl -n hw.ncpu) instead of $(nproc) on macOS

### PR DESCRIPTION
**Pull request by kou**

### CODE REVIEW

**Review by nealrichardson — APPROVED** [review]
I checked for other places where `nproc` is called and they all have this logic around them (though they look at `uname` to determine OS), all except `cpp/examples/minimal_build/build.sh`. But that one corresponds to a ubuntu dockerfile so it's probably fine.

**Review by nealrichardson — CHANGES_REQUESTED** [review]
Actually it looks like this doesn't work: all of the C++ test jobs are failing, e.g. https://github.com/apache/arrow/pull/6160/checks?check_run_id=382886347#step:5:1459

Perhaps switch to check uname, like the others?

File: ci/scripts/cpp_test.sh [hunk 1]
```diff
@@ -28,5 +28,10 @@ export PARQUET_TEST_DATA=${source_dir}/submodules/parquet-testing/data
 export LD_LIBRARY_PATH=${ARROW_HOME}/${CMAKE_INSTALL_LIBDIR:-lib}:${LD_LIBRARY_PATH}
 
 pushd ${build_dir}
-ctest --output-on-failure -j$(nproc)
+if type nproc > dev/null 2>&1; then
```
Comment (kou) [inline ci/scripts/cpp_test.sh]: `/` is missing before `dev/null`...

**Comment by kou** [comment]
Anyway, I changed to use `uname`.

**Review by nealrichardson — APPROVED** [review]
Thanks @kou! I'll merge. The test failures are flaky tests, ticketed https://issues.apache.org/jira/browse/ARROW-7551 and https://issues.apache.org/jira/browse/ARROW-7552

**Pull request closed without merging**